osutil/sync_dir.go

```diff
@@ -0,0 +1,100 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// FileState describes the expected content and meta data of a single file.
+type FileState struct {
+	Content []byte
+	Mode    os.FileMode
+}
+
+var errSameState = fmt.Errorf("file state has not changed")
+
+// EnsureDirState ensures that directory content matches expectations.
+//
+// EnsureDirState enumerates all the files in the specified directory that
+// match the provided pattern (glob). Each enumerated file is checked to ensure
+// that the contents, permissions are what is desired. Unexpected files are
+// removed. Missing files are created and differing files are corrected.  Files
+// not matching the pattern are ignored.
+//
+// The content map describes each of the files that are intended to exist in
+// the directory.  Map keys must be file names relative to the directory.
+// Sub-directories in the name are not allowed.
+//
+// The function stops at the first encountered error but reports all of the
+// changes performed so far. Information about the performed changes is
```

Comment:
That said, with this function so general without a concrete example in the documentation at the top of the function, I'm finding it difficult to follow. Ie, I know what we are trying to achieve (I think! :), and so I am looking at how this function can fail, etc with bad arguments, but with things being generalized I see lots of place where a bad invocation can cause issues. Put another way-- we want to implement a specific thing for keeping files in sync, and we know what we want to use it for, but we are making it general for use cases that I don't know about.

If you want to keep it this way, fine, but I think it worth noting that I found it harder to follow and prove to myself that it was correct than I feel it should be.